File: ios/Classes/RTMChannel.m
//
//  RTMChannel.m
//  agora_rtm
//
//  Created by LY on 2019/8/16.
//

#import "RTMChannel.h"

@implementation RTMChannel

- (instancetype) initWithClientIndex:(NSNumber *)clientIndex channelId:(NSString *)channelId messenger:(id)messenger kit:(AgoraRtmKit*)kit {
  self = [super init];
  if (self) {
    _clientIndex = clientIndex;
    _channelId = channelId;
    _channel = [kit createChannelWithId:channelId delegate:self];
    if (nil == _channel) {
      return nil;
    }
    _messenger = messenger;
    NSString *channelName = [NSString stringWithFormat:@"io.agora.rtm.client%@.channel%@", [clientIndex stringValue], channelId];
    _eventChannel = [FlutterEventChannel eventChannelWithName:channelName binaryMessenger:_messenger];
    if (nil == _eventChannel) {
      return nil;
    }
    
    [_eventChannel setStreamHandler:self];
  }
  return self;
}

- (void) dealloc {
  _clientIndex = nil;
  _channelId = nil;
  _channel = nil;
  _eventChannel = nil;
  _messenger = nil;
}

#pragma - FlutterStreamHandler
- (FlutterError * _Nullable)onCancelWithArguments:(id _Nullable)arguments {
  _eventSink = nil;
  return nil;
}

- (FlutterError * _Nullable)onListenWithArguments:(id _Nullable)arguments eventSink:(nonnull FlutterEventSink)events {
  _eventSink = events;
  return nil;
}

- (void) sendChannelEvent:(NSString *)name params:(NSDictionary*)params {
  if (_eventSink) {
    NSMutableDictionary *dict = [[NSMutableDictionary alloc] initWithDictionary:params];
    dict[@"event"] = name;
    _eventSink([dict copy]);
  }
}

#pragma - AgoraRtmChannelDelegate
- (void)channel:(AgoraRtmChannel *)channel memberJoined:(AgoraRtmMember *)member {
  [self sendChannelEvent:@"onMemberJoined" params:@{@"userId": member.userId, @"channelId": member.channelId}];
}

- (void)channel:(AgoraRtmChannel *)channel memberLeft:(AgoraRtmMember *)member {
  [self sendChannelEvent:@"onMemberLeft" params:@{@"userId": member.userId, @"channelId": member.channelId}];
}

- (void)channel:(AgoraRtmChannel *)channel messageReceived:(AgoraRtmMessage *)message fromMember:(AgoraRtmMember *)member {
  [self sendChannelEvent:@"onMessageReceived" params:@{@"userId": member.userId, @"channelId": member.channelId, @"message": message.text}];
}

@end

File: ios/Classes/RTMChannel.swift
//
//  RTMChannel.swift
//  agora_rtm
//
//  Created by LXH on 2023/5/31.
//

import Foundation
import Flutter
import AgoraRtmKit

class RTMChannel: NSObject, FlutterStreamHandler, AgoraRtmChannelDelegate {
    var eventChannel: FlutterEventChannel
    var eventSink: FlutterEventSink?
    
    init(_ clientIndex: Int, _ channelId: String, _ messenger: FlutterBinaryMessenger) {
        eventChannel = FlutterEventChannel(name: "io.agora.rtm.client\(clientIndex).channel\(channelId)", binaryMessenger: messenger)
        super.init()
        eventChannel.setStreamHandler(self)
    }
    
    func sendEvent(eventName: String, params: Dictionary<String, Any?>) {
        let event: [String: Any?] = [
            "event": eventName,
            "data": params,
        ]
        
        eventSink?(event)
    }
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
    
    func channel(_ channel: AgoraRtmChannel, memberCount count: Int32) {
        sendEvent(eventName: "onMemberCountUpdated", params: [
            "memberCount": count,
        ])
    }
    
    func channel(_ channel: AgoraRtmChannel, attributeUpdate attributes: [AgoraRtmChannelAttribute]) {
        sendEvent(eventName: "onAttributesUpdated", params: [
            "attributeList": attributes.map { $0.toJson() },
        ])
    }
    
    func channel(_ channel: AgoraRtmChannel, messageReceived message: AgoraRtmMessage, from member: AgoraRtmMember) {
        sendEvent(eventName: "onMessageReceived", params: [
            "message": message.toJson(),
            "fromMember": member.toJson(),
        ])
    }
    
    func channel(_ channel: AgoraRtmChannel, memberJoined member: AgoraRtmMember) {
        sendEvent(eventName: "onMemberJoined", params: [
            "member": member.toJson(),
        ])
    }
    
    func channel(_ channel: AgoraRtmChannel, memberLeft member: AgoraRtmMember) {
        sendEvent(eventName: "onMemberLeft", params: [
            "member": member.toJson(),
        ])
    }
}
